File: python/src/day7/main.py
from src.common.file_utils import get_path, read_lines
import re


def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    bags = {}
    for line in lines:
        outer, rest = line.split("contain")
        outer = outer.strip().rstrip("s")

        if rest == " no other bags.":
            bags[outer] = []

        else:
            bags[outer] = []
            contains = rest.split(",")
            for c in contains:
                c = c.strip().replace(".", "")
                r = re.search(r'(\d+) (\S+.*)', c)
                if r:
                    num, i = r.groups()
                    bags[outer].append(i.rstrip("s"))

    def check_bags(bag):
        if bag == "shiny gold bag":
            return True
        if bag in bags:
            return any([check_bags(b) for b in bags[bag]])
        return False

    total_gold = 0
    for b in bags:
        if check_bags(b):
            total_gold += 1

    return total_gold - 1


def part_two(filename: str) -> int:

    lines = read_lines(get_path(__file__, filename))
    bags = {}
    for line in lines:
        outer, rest = line.split("contain")
        outer = outer.strip().rstrip("s")

        if rest == " no other bags.":
            bags[outer] = []

        else:
            bags[outer] = []
            contains = rest.split(",")
            for c in contains:
                c = c.strip().replace(".", "")
                r = re.search(r'(\d+) (\S+.*)', c)
                if r:
                    num, i = r.groups()
                    bags[outer].append((int(num), i.rstrip("s")))

    def count_bags(bag):
        if bag not in bags or bags[bag] == []:
            return 0

        s = 0
        for b in bags[bag]:
            s += b[0] + (b[0] * count_bags(b[1]))
        return s

    return count_bags("shiny gold bag")


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day7/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	lines := input.ReadLines(filename)
	bags := CreateGraph(lines)

	var goldBagInside func(string)bool

	goldBagInside = func(startBag string) bool{
		if contents, bagExists := bags[startBag]; !bagExists || len(contents) == 0 {
			return false
		}
		if startBag == "shiny gold bag"{
			return true
		}
		contents := bags[startBag]
		for _, b := range contents {
			if goldBagInside(b.name){
				return true
			}
		}
		return false
	}

	var s int
	for bag, _ := range(bags) {
		if bag != "shiny gold bag" && goldBagInside(bag){
			s += 1
		}
	}
	return s
}

func Part2(filename string) int {
	lines := input.ReadLines(filename)
	graph := CreateGraph(lines)

	var countBags func(string) int
	
	countBags = func(bag string) int {
		if contents, bagExists := graph[bag]; !bagExists || len(contents) == 0 {
			return 0
		}
		contents, s := graph[bag], 0
		for _, b := range(contents){
			s += b.num + (b.num * countBags(b.name))
		}
		return s
	}
	return countBags("shiny gold bag")
}

func CreateGraph(lines []string) map[string][]innerBag {
	bags := map[string][]innerBag{}
	for _, line := range lines {
		splittedString := strings.Split(line, "contain")
		outerBag, innerBags := splittedString[0], splittedString[1]
		outerBag = strings.TrimSpace(outerBag)
		outerBag = strings.TrimRight(outerBag, "s")
		bags[outerBag] = []innerBag{}
		if innerBags != "no other bags." {
			innerBags = strings.TrimRight(innerBags, ".")
			splitBags := strings.Split(innerBags, ",")
			for _, bag := range splitBags {
				bag = strings.TrimRight(bag, "s")
				re := regexp.MustCompile(`(\d+) (\S+.*)`)
				parts := re.FindStringSubmatch(bag)
				if len(parts) == 3 {
					num, bagName := toInt(parts[1]), parts[2]
					bags[outerBag] = append(bags[outerBag], innerBag{bagName, num})
				}
			}
		}
	}
	return bags
}

type innerBag struct {
	name string
	num  int
}

func sum(arr []int) int {
	var s int
	for _, a := range arr {
		s += a
	}
	return s
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}

func check(err error) {
	if err != nil {
		panic(err)
	}
} 
